### ISSUE TITLE
Read Arabic Numerals (AEN) and Output English Numerals (AWN)

### ISSUE THREAD

**Issue by m0beenali** [tesseract-ocr/tesseract#2791]
### Environment

* **Tesseract Version**: 5.0.0-alpha / 4.0.0
* **Platform**: Ubuntu / macOS Catalina

### Current Behavior: 
Arabic Numerals are read and written in right to left order

### Expected Behavior: 
Reads Arabic numerals and writes those Numerals in English, i.e. 
**Reads** 'ارقام ١٢٣٤٥' and **Outputs** '12345 ارقام'

So, here's what i did,

**1.** I ran text2image with my training_text file
```text2image --text /home/mobeen/customtrain/langdata/ara/ara.training_text 
--outputbase /home/mobeen/customtrain/tiff-box/ara.Arial
--fonts_dir /home/mobeen/Documents/fonts
--font 'Arial'
```
By this, i got tiff and box files as output. I removed the box file created by text2image as it is not in lstm format

**2.** Then I ran
```tesseract /home/mobeen/customtrain/tiff-box/ara.Arial.tif /home/mobeen/customtrain/tiff-box/ara.Arial -l ara-new lstmbox```
this gave me the lstm format box file.

**3.** Next I opened this box file replaced all AEN with AWN and save the file.
Then i ran tesstrain using --my_boxtiff_dir argument, as follows:
```
src/training/tesstrain.sh \
--fonts_dir /home/mobeen/Documents/fonts \
--lang ara --linedata_only --noextract_font_properties \
--langdata_dir ../langdata \
--tessdata_dir ./tessdata \
--output_dir ~/customtrain/aratrain \
--fontlist 'Arial' \
--my_boxtiff_dir /home/mobeen/customtrain/tiff-box
```
this generated the lstmf file and gave me a starter traineddata file.

**5.** Next i ran,
```training/lstmtraining --debug_interval -1 \
--traineddata ~/customtrain/aratrain/ara/ara.traineddata \
--net_spec '[1,36,0,1 Ct3,3,16 Mp3,3 Lfys48 Lfx96 Lrx96 Lfx256 O1c111]' \
--model_output ~/customtrain/araoutput/base --learning_rate 20e-4 \
--train_listfile ~/customtrain/aratrain/ara.training_files.txt \
--eval_listfile ~/customtrain/araeval/ara.training_files.txt \
--max_iterations 3600 &>~/customtrain/araoutput/basetrain.log
In another tereminal window i ran,
tail -f ~/customtrain/araoutput/basetrain.log ```
Wich displayed this: 

``` File /home/mobeen/customtrain/aratrain/ara.Arial.exp0.lstmf line 3 :
Mean rms=0.585%, delta=0.957%, train=2.68%(4.53%), skip ratio=0%
Iteration 3588: GROUND  TRUTH : يف نأ ةفاضإ ١ مالفا و امك خيرات ٢ ةيسيئرلا ٣ مقر ٤ برعلا
Iteration 3588: BEST OCR TEXT : يف نأ ةفشإ ١ مالا و امك خيراا ٢ ةيسيئرلا ٣ مقر ٤ برملا
File /home/mobeen/customtrain/aratrain/ara.Arial.exp0.lstmf line 4 :
Mean rms=0.588%, delta=0.963%, train=2.691%(4.558%), skip ratio=0%
Iteration 3589: GROUND  TRUTH : ىدتنم ٨ نآلا دق ٥ مسق ٧ ةفاضإ _ ٦ عيقوتلا ٩ ةيبرعلا ىدتنم
Iteration 3589: BEST OCR TEXT : ىدتنم ٥ نآلا هق ٥ مسا ٧ ةفاضإ _ ٦ عيقوتلا ٢ ةيبرعلا ىدتنم
File /home/mobeen/customtrain/aratrain/ara.Arial.exp0.lstmf line 5 :
Mean rms=0.59%, delta=0.968%, train=2.705%(4.587%), skip ratio=0%
Iteration 3590: GROUND  TRUTH : ةيزمرلا ٦ ىلإ ٩ جماربلا ٨ ذنم ٥ ١ ىدتنملا ٧ نع ىدتنم
Iteration 3590: BEST OCR TEXT : ةيزمرلا ١ ىلإ ٩ جماربلا ٨ انم ٥ ١ ىدتنسلا ٧ نع ىدتنم
File /home/mobeen/customtrain/aratrain/ara.Arial.exp0.lstmf line 6 :
Mean rms=0.592%, delta=0.971%, train=2.717%(4.61%), skip ratio=0%
Iteration 3591: GROUND  TRUTH : هيف ٧ دمحأ ٩ ةيزمرلا ٣ دوك ٥ رورملا ١ حب هل ٦ ةفاك ٨ ماعلا ٣ يلع
Iteration 3591: BEST OCR TEXT : هيف ٧ دمحأ ٣ ةيزمرلا ٣ دوك ٥ رورملا ٠ نب هل ٦ ةفا ٥ مسقا ٣ يلع
File /home/mobeen/customtrain/aratrain/ara.Arial.exp0.lstmf line 7 :
Mean rms=0.594%, delta=0.976%, train=2.738%(4.643%), skip ratio=0%
Iteration 3592: GROUND  TRUTH : ىلعو ٧ نب ٦ ةكراشملا ٥ خيرات ٨ عيطتست ٩ ىلعألا
Iteration 3592: BEST OCR TEXT : ىلاو ٧ نب ٩ ةكراشملا ٥ خيرقت ٨ عيقطتست ٩ ىلعأل
File /home/mobeen/customtrain/aratrain/ara.Arial.exp0.lstmf line 8 :
Mean rms=0.596%, delta=0.979%, train=2.751%(4.689%), skip ratio=0%
Iteration 3593: GROUND  TRUTH : هيلع ٨ دئاصق ٦ لئاسرلا ٧ برغملا ٥ نيطسلف ١ يه ٣ ماظنلا ٩ تاكراشم
Iteration 3593: BEST OCR TEXT : هيلع ٨ دئاضق ٩ لئاسرلا ٧ برتملا ٥ نيطسلفا ٢ يه ٣ ماظنلا ٩ تاكراشم
File /home/mobeen/customtrain/aratrain/ara.Arial.exp0.lstmf line 9 :
Mean rms=0.599%, delta=0.984%, train=2.765%(4.722%), skip ratio=0%
Iteration 3594: GROUND  TRUTH : / ٩ ةديدج ٦ يذلا نإ ال ٧ سلجم ٩ هب ٠ ىلوألا ٥ روصلا ٨ لا راوزلا
Iteration 3594: BEST OCR TEXT : / ٩ ةديدج ٦ يذلا نإ ال ٧ سدجم ٩ هب ٠ ىلوألا ٨ روصلا ٨ لا راولا
File /home/mobeen/customtrain/aratrain/ara.Arial.exp0.lstmf line 10 :
Mean rms=0.601%, delta=0.987%, train=2.773%(4.739%), skip ratio=0%
Iteration 3595: GROUND  TRUTH : عيضاوم ٨ تاكراشم ٥ انب ٣ تانب ٧ رابخأ ٠ ىلع ٦ ريغ اذه دقو لكشب ٩
Iteration 3595: BEST OCR TEXT : عيضاوم ٨ تاكراشم ٥ انب ٣ تانب ٧ رايخأ ٠ ىلع ٦ ريغ اذه دقو لكشب ٩
File /home/mobeen/customtrain/aratrain/ara.Arial.exp0.lstmf line 11 :
Mean rms=0.602%, delta=0.988%, train=2.777%(4.744%), skip ratio=0%
Iteration 3596: GROUND  TRUTH : خيشلا ٩ ثحبلا ٨ رييغت ٦ نيب ١ مسا ءزجلا ٧ يف لالخ ٥ عوضوملا
Iteration 3596: BEST OCR TEXT : خيللا ٩ ثحبلا ٨ ريغت ٦ نيب ١ مسا ءزجلا ٧ يف لالخ ٥ عوضوملا
File /home/mobeen/customtrain/aratrain/ara.Arial.exp0.lstmf line 12 :
Mean rms=0.603%, delta=0.99%, train=2.782%(4.758%), skip ratio=0%
Iteration 3597: GROUND  TRUTH : موي ٦ نوكي نم ٨ ةيزم١ رلا ٥ىتح ٩ جمارب ٣ زكرم ٧ نأ ٠ عقوملا ريغ
Iteration 3597: BEST OCR TEXT : موي ٦ نوكج نم ٨ ةيزم١ رلا ٥وغح ٦ جمارب ٣ زكرم ٧ نأ ٠ عقوملا ريغ
File /home/mobeen/customtrain/aratrain/ara.Arial.exp0.lstmf line 13 :
Mean rms=0.605%, delta=0.993%, train=2.794%(4.775%), skip ratio=0%
Iteration 3598: GROUND  TRUTH : نم غلبي ٢ نودجاوتملا ٣ ةدهاشم ١ ظفح ٤ تاكراشملا ٠ ةطساوب
Iteration 3598: BEST OCR TEXT : ني علبيب ٣ نوضجاوتملا ٣ ةداضشم ١ ثفنح ٤ تاكراشملا ٠ ةطساوب
File /home/mobeen/customtrain/aratrain/ara.Arial.exp0.lstmf line 14 :
Mean rms=0.608%, delta=1%, train=2.819%(4.825%), skip ratio=0%
Iteration 3599: GROUND  TRUTH : يصخشلا ٨ دمحم ٥ ءاوح ١ جمارب هل ٦ ةروصلا و ٧ ماظن ٩ ماع ناكو
Iteration 3599: BEST OCR TEXT : يصخشلا ٨ دمحم ٥ ءاوح ١ جمارب هل ١ ةروصلا و ٧ ماظنن ٩ ماع نقكر
File /home/mobeen/customtrain/aratrain/ara.Arial.exp0.lstmf line 15 :
Mean rms=0.61%, delta=1.002%, train=2.831%(4.844%), skip ratio=0%
At iteration 2182/3600/3600, Mean rms=0.61%, delta=1.002%, char train=2.831%, word train=4.844%, skip ratio=0%,  New worst char error = 2.831 wrote checkpoint.

Finished! Error rate = 0.064
```
As you can see it still reads AEN as AEN not AWN

**Comment by Shreeshrii** [comment]
> I removed the box file created by text2image as it is not in lstm format

`text2image `produces box files that can be used for `lstmtraining`

`tesseract `with `makebox `for scanned images does not produce box files compatible for `lstmtraining`.

**Comment by m0beenali** [comment]
The rest of the process is correct? Please it would be really helpful....

**Comment by Shreeshrii** [comment]
I  am testing. WIll post when I have something definite.

**Comment by m0beenali** [comment]
ok, Thanks @Shreeshrii

**Comment by Shreeshrii** [comment]
Please test with the attached two traineddata files. 

[aranumberLayer_1.57_29554_53100.zip](https://github.com/tesseract-ocr/tesseract/files/3926150/aranumberLayer_1.57_29554_53100.zip)

Training data sample  is in following files.
[araNumber.zip](https://github.com/tesseract-ocr/tesseract/files/3926190/araNumber.zip)



Sample from training log
```
Iteration 75283: GROUND  TRUTH : دق كدبعاذانأاه » كيبل « ىنعمو : لاقو ، )8(ليلخلا كلذ لاق ، ناكملاب تببلأ
Iteration 75283: BEST OCR TEXT : دق كدبعاذاأاه » كيبل « ىنعمو : لاق ، )8(ليلخلا كلذ لاق ، ناكملاب تببلأ
File /home/ubuntu/OCR_GS_Data/ara/AWN/book_IbnQutayba.Adab_7_final_200_000144.lstmf line 0 :
Mean rms=0.46%, delta=0.258%, train=2.535%(3.994%), skip ratio=0%
Iteration 75284: GROUND  TRUTH : . ةيمجعلا مهف ءوس ىلعو ،ةيبلقصلا )5(ةهالب ىلعو ، هتزيرغ ةوابغ
File /home/ubuntu/OCR_GS_Data/ara/AWN/book_Jahiz.Hayawan_7_final_000833.lstmf line 0 (Perfect):
Mean rms=0.46%, delta=0.258%, train=2.535%(3.994%), skip ratio=0%
Iteration 75285: GROUND  TRUTH : امسا الإ لاعفتلا )4(مالكلا يف نوكي الو ، )3(اراركتو اريركت هتررك : لثم
File /home/ubuntu/OCR_GS_Data/ara/AWN/book_IbnQutayba.Adab_7_final_000693.lstmf line 0 (Perfect):
Mean rms=0.459%, delta=0.258%, train=2.524%(3.961%), skip ratio=0%
Iteration 75286: GROUND  TRUTH : عقربلا « و ، )2(لصألا » رصنعلاو رصنعلا « و ، يربلا لصبلل » لصنعو
File /home/ubuntu/OCR_GS_Data/ara/AWN/book_IbnQutayba.Adab_7_final_200_000440.lstmf line 0 (Perfect):
Mean rms=0.46%, delta=0.258%, train=2.524%(3.961%), skip ratio=0%
Iteration 75287: GROUND  TRUTH : .)3(ءامسلا نم هللا ينجوزو نكولهأ نكجوز :لوقتو هئاسن
File /home/ubuntu/OCR_GS_Data/ara/AWN/book_IbnAthir.Kamil_7_final_200_000519.lstmf line 0 (Perfect):
Mean rms=0.459%, delta=0.258%, train=2.524%(3.961%), skip ratio=0%
Iteration 75288: GROUND  TRUTH : ىحو : لثم ، يلح عمج يلحلا : ءارفلا لاق . » ىلحلا « :)6(تلقف
Iteration 75288: BEST OCR TEXT : يحو : لثم ، يلح عمج يلحلا : ءارفلا لاق . » يلحلا « : )6(تلقف
File /home/ubuntu/OCR_GS_Data/ara/AWN/book_IbnQutayba.Adab_7_final_000082.lstmf line 0 :
Mean rms=0.46%, delta=0.259%, train=2.533%(3.981%), skip ratio=0%
Iteration 75289: GROUND  TRUTH : سلاجم يفو ، كلسي مل ةمثألا ليبس يف 1... لجرل ىبوط : لاق روبزلا أرق اذإ
File /home/ubuntu/OCR_GS_Data/ara/AWN/book_Yacqubi.Tarikh_7_final_b_200_000750.lstmf line 0 (Perfect):
Mean rms=0.46%, delta=0.259%, train=2.532%(3.974%), skip ratio=0%
Iteration 75290: GROUND  TRUTH : : انلق : لاق . )4(هورقت ال يعاودلاو ، هورعت ال بابلا اذه يف سواسولا
File /home/ubuntu/OCR_GS_Data/ara/AWN/book_Jahiz.Hayawan_7_final_b_200_000568.lstmf line 0 (Perfect):
Mean rms=0.46%, delta=0.259%, train=2.532%(3.974%), skip ratio=0%
Iteration 75291: GROUND  TRUTH : . ةدحولا نم ملسأ الو ، باتك نم )1( عنمأ الو ، ربق نم ظعوأ رأ مل
File /home/ubuntu/OCR_GS_Data/ara/AWN/book_Jahiz.Hayawan_7_final_000761.lstmf line 0 (Perfect):
Mean rms=0.46%, delta=0.259%, train=2.532%(3.974%), skip ratio=0%
Iteration 75292: GROUND  TRUTH : دق كدبعاذانأاه » كيبل « ىنعمو : لاقو ، )8(ليلخلا كلذ لاق ، ناكملاب تببلأ
Iteration 75292: BEST OCR TEXT : دق كدبعاذنأاه » كيبل « ىنعمو : لاقو ، )8(ليلخلا كلذ لاقو ، ناكملاب تببلأ
File /home/ubuntu/OCR_GS_Data/ara/AWN/book_IbnQutayba.Adab_7_final_a_000144.lstmf line 0 :
Mean rms=0.461%, delta=0.259%, train=2.534%(3.988%), skip ratio=0%
Iteration 75293: GROUND  TRUTH : : اولاق ، بابلا اذه ضعب يف » لعف « ىلع » ليعف « )6(لخدي دقو
File /home/ubuntu/OCR_GS_Data/ara/AWN/book_IbnQutayba.Adab_7_final_000161.lstmf line 0 (Perfect):
Mean rms=0.461%, delta=0.259%, train=2.534%(3.988%), skip ratio=0%
Iteration 75294: GROUND  TRUTH : انركذ دقو ،)5(رما وذ اهل لاقي رامنأ ةوزغ ،ملعص ،هللا لوسر ازغ اهيفو :يدقاولا لاقو
File /home/ubuntu/OCR_GS_Data/ara/AWN/book_IbnAthir.Kamil_7_final_000118.lstmf line 0 (Perfect):
Mean rms=0.461%, delta=0.259%, train=2.534%(3.988%), skip ratio=0%
Iteration 75295: GROUND  TRUTH : هتلقلق « : لوقتو ، ردصم وهف هترسك اذإف ؛ مسا )2(حوتفم وهو ، كلذ
File /home/ubuntu/OCR_GS_Data/ara/AWN/book_IbnQutayba.Adab_7_final_200_000106.lstmf line 0 (Perfect):
```

**Comment by m0beenali** [comment]
Thanks alot @Shreeshrii. I've tested both traineddata files numbers are read in arabic and outputed in english and I've also learned alot by checking your trianing data sample files.

Test results were very good in reading numerals but character recognition is not much accurate sometimes, so should I finetune any of the traineddata files for increasing accuracy for alphabets?

**Comment by Shreeshrii** [comment]
The focus of this test was just for the Arabic numbers, for AEN to AWN. This required replacement of top layer. You will need a much larger training text.

**Comment by Shreeshrii** [comment]
Please check the attached zip with traineddata files for Arabic characters and AEN and compare with tessdata_best/ara.

[araOpenITILayer_1.167_262370_480400.zip](https://github.com/tesseract-ocr/tesseract/files/3927249/araOpenITILayer_1.167_262370_480400.zip)

**Comment by m0beenali** [comment]
Thanks alot! again @Shreeshrii i will check these files and update soon

**Comment by m0beenali** [comment]
@Shreeshrii can you write the steps you followed to train this data using the training files you provided in above comment. Thanks a lot again!

**Comment by anuradhab2018** [comment]
> Please check the attached zip with traineddata files for Arabic characters and AEN and compare with tessdata_best/ara.
> 
> [araOpenITILayer_1.167_262370_480400.zip](https://github.com/tesseract-ocr/tesseract/files/3927249/araOpenITILayer_1.167_262370_480400.zip)

hi Shree...I have a small question here...Do I have to replace the trained.ara file in the tessdata folder,with the above trained data files for extracting the arabic numbers in windows?plz help

**Comment by wollmers** [comment]
IMHO somebody should close this issue.
